smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,475 @@
+
+        Parameters
+        ----------
+        min_budget : float
+            minimum budget allowed for 1 run of successive halving
+        max_budget : float
+            maximum budget allowed for 1 run of successive halving
+        eta : float
+            'halving' factor after each iteration in a successive halving run
+        init_chal : int
+            number of challengers to consider for the initial budget
+        """
+
+        if eta <= 1:
+            raise ValueError('eta must be greater than 1')
+        self.eta = eta
+
+        # BUDGETS
+
+        # - if only 1 instance was provided & quality objective, then use cutoff as budget
+        # - else, use instances as budget
+        if not self.run_obj_time and len(self.instances) <= 1:
+            # budget with cutoff
+            # cannot run successive halving with cutoff as budget if budget limits are not provided!
+            if min_budget is None or \
+                    (max_budget is None and self.cutoff is None):
+                raise ValueError("Successive Halving with runtime-cutoff as budget (i.e., only 1 instance) "
+                                 "requires parameters min_budget and max_budget/cutoff for intensification!")
+
+            # if both cutoff and max_budget are provided, then warn user
+            if self.cutoff is not None and max_budget is not None:
+                warnings.warn('Successive Halving with runtime-cutoff as budget: '
+                              'Both max budget (%d) and runtime-cutoff (%d) were provided. Max budget will be used.' %
+                              (self.max_budget, len(self.instances)))
+
+            self.min_budget = min_budget
+            self.max_budget = max_budget if max_budget else self.cutoff
+            self.cutoff_as_budget = True
+
+        else:
+            # budget with instances
+            if self.run_obj_time and len(self.instances) <= 1:
+                warnings.warn("Successive Halving has objective 'runtime' but only 1 instance-seed pair.")
+            self.min_budget = 1 if min_budget is None else int(min_budget)
+            self.max_budget = len(self.instances) if max_budget is None else int(max_budget)
+            self.cutoff_as_budget = False
+
+            # max budget cannot be greater than number of instance-seed pairs
+            if self.max_budget > len(self.instances):
+                raise ValueError('Max budget cannot be greater than the number of instance-seed pairs')
+            if self.max_budget < len(self.instances):
+                warnings.warn('Max budget (%d) does not include all instance-seed pairs (%d)' %
+                              (self.max_budget, len(self.instances)))
+
+        self.logger.debug("Running Successive Halving with '%s' as budget" % self.cutoff_as_budget)
+
+        # number configurations to consider for a full successive halving iteration
+        self.max_sh_iter = np.floor(np.log(self.max_budget / self.min_budget) / np.log(eta))
```

Comment:
Could you please add a reference to the line in Algorithm 1 of [Li et al.](http://jmlr.org/papers/volume18/16-558/16-558.pdf)? This would make it easier to read (I'm actually right now not sure which line this belongs to and which variable you are computing here).

Reply:
Also, please add a comment which variable each of the variables here corresponds to in Algorithms 1 of [Li et al.](http://jmlr.org/papers/volume18/16-558/16-558.pdf).

Reply:
The 'max_sh_iter' variable is just a precomputed variable used for sampling configurations when SuccessiveHalving is used independently.
If configurations come from hyperband, then this variable is not used.

Reply:
The implementation was based on BOHB paper and hpbandster package. I'll add these as reference.

Reply:
Is this `s_max` from Algorithm 1 of the BOHB paper?

Reply:
Could you please add a comment to each of the three variables you pre-compute what they're doing exactly?